### DISCUSSION TITLE
Decoding sx1261 SPI commands in FSK mode

### DISCUSSION THREAD

**Discussion by TimelessNL** [jgromes/RadioLib#483] [Q&A] [answered]
Hi,

Thanks for this great library!

### Introduction:
I'm currently in a somewhat strange situation. I'm trying to decode the SPI communication from a `SX1261`.
The complete story is that I have some LoRaWAN tags around that periodically send out a FSK message using a `SX1261` that I would like to receive. Unfortunately I do only have a [SX1272](https://development.libelium.com/waspmote-technical-guide/lora-module) at my disposal but from what I've read this radio should be capable of receiving the `SX1261` if configured correctly.

### Decoding SPI
`SetRfFrequency` looks like this:
![afbeelding](https://user-images.githubusercontent.com/10838968/154860707-032fda29-3239-4d58-b272-310e3d550eb3.png)
Which to my calculations has to be 868,79Mhz

The `SyncWord` is directly written to a register:
![afbeelding](https://user-images.githubusercontent.com/10838968/154860872-458ca83f-5e93-4b08-a362-29d14687f271.png)
That seems to be 3 bytes `0xC1` `0x94` `0xC1`

The `SetModulationParams` are as follows:
![afbeelding](https://user-images.githubusercontent.com/10838968/154861005-b47125dd-708e-493d-99dc-5160ec41edee.png)
The `bitrate` seems to be 50kbps
`PulseShape` is `0x0B` which would be `Gaussian BT 1` according to the datasheet
`BandWidth` is `0x1F` which would be `RX_BW_4800` as 4.8kHz
`Frequency Deviation` Is probably `24.9`

### Code:
When I flash my NodeMCU (ESP-12E) module with the following sketch:
```
// include the library
#include <RadioLib.h>

SX1272 radio = new Module(D8, D2, D0, RADIOLIB_NC);

void setup() {
  Serial.begin(115200);

  // initialize SX1272 FSK modem with default settings
  Serial.print(F("[SX1272] Initializing ... "));
  int state = radio.beginFSK();
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }

  Serial.print("SX1272 ver: ");
  Serial.println(radio.getChipVersion());

  // Configure radio settings
  state = radio.setFrequency(868.79);
  state = radio.setBitRate(50.0);
  state = radio.setFrequencyDeviation(24.9);
  state = radio.setRxBandwidth(4.8);
  state = radio.setOutputPower(10.0);
  state = radio.setCurrentLimit(100);
  state = radio.setDataShaping(RADIOLIB_SHAPING_1_0);
  uint8_t syncWord[] = {0xC1, 0x94, 0xC1};
  state = radio.setSyncWord(syncWord, 3);
  state = radio.disableAddressFiltering();
  state = radio.setOOK(false);
  state = radio.setCRC(false); 
  if (state != RADIOLIB_ERR_NONE) {
    Serial.print(F("Unable to set configuration, code "));
    Serial.println(state);
    while (true);
  }
}

void loop() {
  // receive FSK packet
  String str;
  int state = radio.receive(str);

  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("[SX1272] Received packet!"));
    Serial.print(F("[SX1272] Data:\t"));
    Serial.println(str);      
  } else if (state == RADIOLIB_ERR_RX_TIMEOUT) {
    Serial.println(F("[SX1272] Timed out while waiting for packet!"));
  } else {
    Serial.println(F("[SX1272] Failed to receive packet, code "));
    Serial.println(state);
  }
}
```

### Result:
The SX1272 only returns timeouts:
```
[SX1272] Initializing ... success!
20:47:11.555 -> SX1272 ver: 34
20:47:11.555 -> [SX1272] Timed out while waiting for packet!
20:47:11.555 -> [SX1272] Timed out while waiting for packet!
20:47:11.590 -> [SX1272] Timed out while waiting for packet!
```

### Problem:
So I'm guessing my reverse SPI -> Code assumptions are somewhat incorrect, maybe someone can enlighten me on what I'm doing wrong?

**Comment by jgromes** [comment, marked as answer]
Going through your SPI decodes, it all seems correct (though the frequency seems more like 868.800 MHz), however, you're missing a crucial piece of the puzzle - the packet format. For example, you know the sync word starts with 0xC194C1, but you don't really know how long it is.

Try to find the `SetPacketParams` transaction (opcode 0x8C), that should give you the remaining information.

I would also use the interrupt-driven reception, since the 50 kbps data rate is quite high and therefore it's likely you will miss it (blocking mode timeout is calculated based on the configured data rate).

Finally, as this is not really an issue (or at least it doesn't seem to be so far), I'm converting this to a discussion.

**Comment by TimelessNL** [comment]
Hi,

Sorry for the late reply, but you're right!
I had to look into the `SetPacketParams` some more, to discover I was missing the whitening parameter.
I'm now able to receive the packets which unfortunately are still encrypted, but the library works so thanks for that!
